src/main/java/com/mixpanel/android/viewcrawler/ViewVisitor.java

```diff
@@ -119,11 +123,35 @@ protected String name() {
         private final Object[] mOriginalValueHolder;
     }
 
+    public static class LayoutUpdateException extends IllegalStateException {
+        public LayoutUpdateException(String message, JSONObject info) {
+            super(message);
+            mErrorInfo = info;
+        }
+
+        public JSONObject getErrorInfo() {
+            return mErrorInfo;
+        }
+
+        private JSONObject mErrorInfo;
+    }
+
     public static class LayoutUpdateVisitor extends ViewVisitor {
-        public LayoutUpdateVisitor(List<Pathfinder.PathElement> path, LayoutRule args) {
+        public LayoutUpdateVisitor(List<Pathfinder.PathElement> path, LayoutRule args, String name) {
             super(path);
             mOriginalValues = new WeakHashMap<View, LayoutRule>();
             mArgs = args;
+            mName = name;
+
+            mHorizontalRules = new ArrayList<>(Arrays.asList(
```

Comment:
This looks like it could be a static, immutable member (and should probably be a set?)